### PR TITLE
Iteration metadata support not properly implemented

### LINKED ISSUE #468: Iteration metadata support not properly implemented
According to https://quarkus.io/guides/qute-reference#loop_section: 

> It’s also possible to access the iteration metadata inside the loop via the following keys:
> - count - 1-based index
> - index - zero-based index
> - hasNext - true if the iteration has more elements
> - isLast - true if hasNext == false
> - isFirst - true if count == 1
> - odd - true if the zero-based index is odd
> - even - true if the zero-based index is even
> - indexParity - outputs odd or even based on the zero-based index value
> 
> However, the keys cannot be used directly. Instead, a prefix is used to avoid possible collisions with variables from the outer scope. By default, the alias of an iterated element suffixed with an underscore is used as a prefix. For example, the hasNext key must be prefixed with it_ inside an {#each} section: {it_hasNext}.
> 
```html
{#each items}
  {it_count}. {it.name} 
  {#if it_hasNext}<br>{/if} 
{/each}
```
Currently, completion offers the raw metadata keys (index, count...) instead of the prefixed flavor (it_count or item_count, depending on the alias), which is incorrect (the app will fail to run). 
And validation fails for the correct prefixed keys:

Screenshot 2021-12-01 at 19 23 18

### PR DESCRIPTION
**Pull request by angelozerr**
Iteration metadata support not properly implemented

Fixes #468

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by fbricon** [comment]
Wondering whether hover should show `count` or `itm_count`:
Screenshot 2021-12-13 at 11 11 16

@mkouba @maxandersen do you have a preference?

**Comment by fbricon** [comment]
renaming de `item` dans `{#for item in items}` doesn't rename `item_count`

**Comment by angelozerr** [comment]
> renaming item in {#for item in items} doesn't rename item_count

fixed

**Comment by fbricon** [comment]
assigned metadata is not linked-edited

Dec-14-2021 15-38-26

**Comment by angelozerr** [comment]
> assigned metadata is not linked-edited

fixed for let parameter value but not for method parameter (it will managed in https://github.com/redhat-developer/quarkus-ls/issues/486)

**Comment by fbricon** [comment]
My template looks like 
```
{#for item in items}
    <h2>{item.name}</h2>    
    {#let foo="bar"}
      <p>{foo}</p>
    {/}
{/}

```

When clicking on `item`, in the #for statement, I get this exception:
```
Caused by: java.lang.NullPointerException: Cannot invoke "com.redhat.qute.parser.expression.ObjectPart.getPartName()" because "objectPart" is null
	at com.redhat.qute.utils.QuteSearchUtils.tryToCollectObjectPartOrParameter(QuteSearchUtils.java:260)
	at com.redhat.qute.utils.QuteSearchUtils.searchReferencedObjects(QuteSearchUtils.java:218)
	at com.redhat.qute.utils.QuteSearchUtils.searchReferencedObjects(QuteSearchUtils.java:248)
	at com.redhat.qute.utils.QuteSearchUtils.searchReferencedObjects(QuteSearchUtils.java:183)
	at com.redhat.qute.utils.QuteSearchUtils.searchReferencedObjects(QuteSearchUtils.java:171)
	at com.redhat.qute.services.QuteLinkedEditing.findLinkedEditingRanges(QuteLinkedEditing.java:49)
	at com.redhat.qute.services.QuteLanguageService.findLinkedEditingRanges(QuteLanguageService.java:151)
	at com.redhat.qute.ls.template.TemplateFileTextDocumentService.lambda$linkedEditingRange$19(TemplateFileTextDocumentService.java:236)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1311)

```

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/QuteSearchUtils.java [hunk 1]
```diff
@@ -206,17 +245,40 @@ private static void searchReferencedObjects(String partName, PartNameMatcher mat
...
+	}
+
+	public static void tryToCollectObjectPartOrParameter(String partName, PartNameMatcher matcher,
+			Expression expression, Node owerNode, BiConsumer<Node, Range> collector) {
+		// Check if the current expression references the given part name
+		if (!tryToCollectObjectPart(partName, matcher, expression, collector)) {
+			if (owerNode.getKind() == NodeKind.Section) {
+				Section onwerSection = (Section) owerNode;
+				// Check the current expression references a metadata of the section
+				ObjectPart objectPart = expression.getObjectPart();
```
Comment (fbricon) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/QuteSearchUtils.java:259]: objectPart can be null, leading to some NPE

**Comment by fbricon** [comment]
quarkus.qute.iteration-metadata-prefix=<alias?> is ignored

**Comment by angelozerr** [comment]
> quarkus.qute.iteration-metadata-prefix=<alias?> is ignored

I created an issue at https://github.com/redhat-developer/quarkus-ls/issues/491

**Comment by angelozerr** [comment]
@fbricon can we merge this PR?

**Comment by fbricon** [comment]
thanks @angelozerr

**Pull request merged**